runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -293,22 +293,29 @@ State waitUntilFinish(
 
   @Override
   public State cancel() throws IOException {
-    Job content = new Job();
-    content.setProjectId(projectId);
-    content.setId(jobId);
-    content.setRequestedState("JOB_STATE_CANCELLED");
-    try {
-      dataflowOptions.getDataflowClient().projects().jobs()
-          .update(projectId, jobId, content)
-          .execute();
-    } catch (IOException e) {
-      String errorMsg = String.format(
-          "Failed to cancel the job, please go to the Developers Console to cancel it manually: %s",
-          MonitoringUtil.getJobMonitoringPageURL(getProjectId(), getJobId()));
-      LOG.warn(errorMsg);
-      throw new IOException(errorMsg, e);
+    State state = getState();
+    if (!state.isTerminal()) {
+      Job content = new Job();
+      content.setProjectId(projectId);
+      content.setId(jobId);
+      content.setRequestedState("JOB_STATE_CANCELLED");
+      try {
+        dataflowOptions.getDataflowClient().projects().jobs()
+            .update(projectId, jobId, content)
+            .execute();
+      } catch (IOException e) {
```

Comment:
Are you able to look at the failure reason instead of checking job state?

This way it closes a race condition where the job reaches a terminal condition between the fetching state and cancel calls